enclave-runner/src/usercalls/mod.rs

```diff
@@ -343,52 +352,234 @@ impl EnclaveState {
+                            }
+                        };
+                        if res.is_err() {
+                            let cmd = enclave_cloned.kind.as_command().unwrap();
+                            let mut cmddata = cmd.data.lock().unwrap();
+                            cmddata.running_secondary_threads -= 1;
+                            if cmddata.running_secondary_threads == 0 {
+                                cmd.wait_secondary_threads.notify_all();
+                            }
+                        }
+                    }
+                });
+            }
+            // ///////////////////////////////
+            // // main syscall polling loop //
+            // ///////////////////////////////
+            'outer: loop {
+                let maybe_block_recv = |block| if block {
+                    io_queue_receive.recv().ok()
+                } else {
+                    io_queue_receive.try_recv().ok()
+                };
+
+                'inner: while let Some((coresult, mut state, mode, buf)) = maybe_block_recv(true) {
+
+                    match coresult {
+                        CoResult::Return((tcs,v1, v2)) => {
+                            if mode == EnclaveEntry::ExecutableMain {
+                                // return and do error checking
+                                main_return = Err(EnclaveAbort::MainReturned);
+                                break 'outer;
+                            }
+
+                            assert_eq!(
+                                (v1, v2),
+                                (0, 0),
+                                "Expected enclave thread entrypoint to return zero"
+                            );
+                            let cmd = state.enclave.kind.as_command().unwrap();
+                            let mut cmddata = cmd.data.lock().unwrap();
+                            cmddata.running_secondary_threads -= 1;
+                            if cmddata.running_secondary_threads == 0 {
+                                cmd.wait_secondary_threads.notify_all();
+                            }
+                            // If the enclave is in the exit-state, threads are no
+                            // longer able to be launched
+                            if !state.enclave.exiting.load(Ordering::SeqCst) {
+                                cmddata.threads_queue.push(StoppedTcs {
+                                    tcs,
+                                    event_queue: state.event_queue,
+                                });
+                            }
+                        },
+                        CoResult::Yield(usercall) => {
+                            let result;
+                            {
+                                let mut on_usercall =
+                                    |p1, p2, p3, p4, p5| dispatch(&mut Handler(&mut state, Some(&work_sender)), p1, p2, p3, p4, p5);
```

Comment:
EnclaveState is shared between threads, the work sender doesn't need to be shared between threads, it's only used by the usercall thread. Also, after the main enclave thread exits, the work_sender needs to be dropped. If I keep it in the EnclaveState then I'll have to jump through hoops to make this happen (if it can happen) and I'm unsure if it is worth it.

Reply:
The current setup kind of breaks the abstraction layer (`Handler`), let me think if there's a better way.